content/posts/building_jnigen.md

```diff
@@ -0,0 +1,262 @@
+
+## Configuration
+
+My initial plan was to expose `jnigen` as an API, so that users can write a dart script in `tool/` directory of the project and call it. However, Daco suggested to provide a YAML configuration similar to what `ffigen` does. It was indeed a better idea from end-user perspective. (At least, this project is not kubernetes-scale and will never require templated YAML packed in an OCI container image).
+
+Another nice effect of implementing YAML config is that it steered the design of configuration towards being more _declarative_ and tidy. In my initial design, user would've to imperatively call the maven utilities described above, using an API. Later, I just made it another configuration parameter, which is much nicer.
+
+## Other stuff
+
+Code generation is heavily used even inside the project. One set of tests which test Dart+C bindings are replicated to test pure dart bindings. This can't be abstracted in code because imports will be different, so code generation is the practical option here. Similarly, a script written against `ffigen`'s internal AST representation is used to generate some C wrappers from `jni.h` of Android NDK.
+
+CheckJNI on Android has been quite helpful. In the beginning, some functionality was only tested on standalone using unit tests. We wanted to run same unit tests on Android. Since there's no provision in flutter for running Unit tests on another device, we made a stub application which ran these tests using integration test mechanism. This uncovered bugs which didn't occur on standalone target. For instance the spec defines `NewGlobalRef` can't be called when an exception is pending. However it happened to work on standalone Dart target (which uses OpenJDK using the JNI invocation API), and the bug was found when the same test was run on Android.
+
+Performance on small tests seems good so far. On a synthetic benchmark which mainly measured trivial calls, we can see more than 10x average improvement over method channels. In practice, if you call into Java, Java code needs to justify it by doing some heavy-lifting anyway. So I believe the main value of the code-generation interop is ergonomics, and then performance.
+
+[Some rudimentary benchmarks measuring the overhead against platform channels](/images/building_jnigen/rudimentary_benchmarks.png)
+
+## Future work
+
+I have to admit that I underestimated the complexity of this project. It's usable in current state, but there's lot of work to be done. Having almost no prior experience with real world programming, I am perplexed at the architectural detail it took to get the smallest things working.
```

Comment:
"current state" was in December? Maybe clarify.

The current state at time of writing is that we advertise it in Google I/O 🎉